libsolidity/interface/CompilerStack.cpp

```diff
@@ -369,26 +376,44 @@ void CompilerStack::resolveImports()
 		for (ASTPointer<ASTNode> const& node: _source->ast->nodes())
 			if (ImportDirective const* import = dynamic_cast<ImportDirective*>(node.get()))
 			{
-				string const& id = import->identifier();
-				if (!m_sources.count(id))
+				string path = absolutePath(import->identifier(), _sourceName);
+				import->annotation().absolutePath = path;
+				if (!m_sources.count(path))
 					BOOST_THROW_EXCEPTION(
 						Error(Error::Type::ParserError)
 							<< errinfo_sourceLocation(import->location())
 							<< errinfo_comment("Source not found.")
 					);
 
-				toposort(&m_sources[id]);
+				toposort(path, &m_sources[path]);
 			}
 		sourceOrder.push_back(_source);
 	};
 
 	for (auto const& sourcePair: m_sources)
 		if (!sourcePair.second.isLibrary)
-			toposort(&sourcePair.second);
+			toposort(sourcePair.first, &sourcePair.second);
 
 	swap(m_sourceOrder, sourceOrder);
 }
 
+string CompilerStack::absolutePath(string const& _path, string const& _reference) const
+{
+	// Anything that does not start with `.` is an absolute path.
+	if (_path.empty() || _path.front() != '.')
+		return _path;
+	using path = boost::filesystem::path;
+	path p(_path);
+	path result(_reference);
+	result.remove_filename();
+	for (path::iterator it = p.begin(); it != p.end(); ++it)
```

Comment:
`path::make_absolute` does not remove `.` or `..` entries. `path::canonical` does, but it also checks for symlinks in the actual filesystem, which we do not want to have (because the path does not need to have any relation to the filesystem the compiler is currently running on), so this hybrid approach was taken.